File: daemon/internal/setupwizard/permissions.go
package setupwizard

import "github.com/dopejs/dope-agent/daemon/internal/identity"

const (
	PermissionSecretsManage      = string(identity.PermissionSecretsManage)
	PermissionIntegrationsManage = string(identity.PermissionIntegrationsManage)
	PermissionCredentialsInspect = string(identity.PermissionCredentialsInspect)
)

func CanMutateSetup(tc identity.TenantContext) bool {
	return tc.TenantID != "" &&
		tc.PrincipalID != "" &&
		identity.HasPermission(tc.Permissions, identity.PermissionSecretsManage) &&
		identity.HasPermission(tc.Permissions, identity.PermissionIntegrationsManage)
}

func CanInspectSetup(tc identity.TenantContext) bool {
	return tc.TenantID != "" &&
		tc.PrincipalID != "" &&
		identity.HasPermission(tc.Permissions, identity.PermissionCredentialsInspect)
}

func RequireMutation(tc identity.TenantContext) error {
	if !CanMutateSetup(tc) {
		return ErrPermissionDenied
	}
	return nil
}

func RequireInspection(tc identity.TenantContext) error {
	if !CanInspectSetup(tc) {
		return ErrPermissionDenied
	}
	return nil
}

File: rs/setupwizard/src/permissions.rs
//! Setup permission checks (port of `permissions.go`).

use dope_identity::{Permission, TenantContext, has_permission};

use crate::types::SetupError;

pub const PERMISSION_SECRETS_MANAGE: &str = "secrets.manage";
pub const PERMISSION_INTEGRATIONS_MANAGE: &str = "integrations.manage";

#[must_use]
pub fn can_mutate_setup(tc: &TenantContext) -> bool {
    !tc.tenant_id.is_empty()
        && !tc.principal_id.is_empty()
        && has_permission(&tc.permissions, Permission::SecretsManage)
        && has_permission(&tc.permissions, Permission::IntegrationsManage)
}

#[must_use]
pub fn can_inspect_setup(tc: &TenantContext) -> bool {
    !tc.tenant_id.is_empty()
        && !tc.principal_id.is_empty()
        && has_permission(&tc.permissions, Permission::CredentialsInspect)
}

pub fn require_mutation(tc: &TenantContext) -> Result<(), SetupError> {
    if !can_mutate_setup(tc) {
        return Err(SetupError::PermissionDenied);
    }
    Ok(())
}

pub fn require_inspection(tc: &TenantContext) -> Result<(), SetupError> {
    if !can_inspect_setup(tc) {
        return Err(SetupError::PermissionDenied);
    }
    Ok(())
}
